cpp/src/io/statistics/column_stats.cu

```diff
@@ -0,0 +1,760 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "column_stats.h"
+#include <math_constants.h>
+#include <io/utilities/block_utils.cuh>
+
+namespace cudf {
+namespace io {
+
+struct stats_state_s
+{
+    stats_column_desc col;
+    statistics_group group;
+    statistics_chunk ck;
+    volatile statistics_val warp_min[32];
+    volatile statistics_val warp_max[32];
+    volatile statistics_val warp_sum[32];
+};
+
+struct merge_state_s
+{
+    stats_column_desc col;
+    statistics_merge_group group;
+    statistics_chunk ck;
+    volatile statistics_val warp_min[32];
+    volatile statistics_val warp_max[32];
+    volatile statistics_val warp_sum[32];
+    volatile uint32_t warp_non_nulls[32];
+    volatile uint32_t warp_nulls[32];
+};
+
+
+inline __device__ int64_t WarpReduceMinInt(int64_t vmin)
+{
+    int64_t v = SHFL_XOR(vmin, 1);
+    vmin = min(vmin, v);
+    v = SHFL_XOR(vmin, 2);
+    vmin = min(vmin, v);
+    v = SHFL_XOR(vmin, 4);
+    vmin = min(vmin, v);
+    v = SHFL_XOR(vmin, 8);
+    vmin = min(vmin, v);
+    v = SHFL_XOR(vmin, 16);
+    return min(vmin, v);
+}
+
+inline __device__ int64_t WarpReduceMaxInt(int64_t vmax)
+{
+    int64_t v = SHFL_XOR(vmax, 1);
+    vmax = max(vmax, v);
+    v = SHFL_XOR(vmax, 2);
+    vmax = max(vmax, v);
+    v = SHFL_XOR(vmax, 4);
+    vmax = max(vmax, v);
+    v = SHFL_XOR(vmax, 8);
+    vmax = max(vmax, v);
+    v = SHFL_XOR(vmax, 16);
+    return max(vmax, v);
+}
+
+inline __device__ double WarpReduceMinFloat(double vmin)
+{
+    double v = SHFL_XOR(vmin, 1);
+    vmin = (v < vmin) ? v : vmin;
```

Comment:
I tried to follow what parquet-cpp does (exclude NaNs from the min/max range). Did not know about the -0 and +0 issue. If -0 is less than +0 in the float comparison then we should be ok.
If not, I'll have to presumably change a +0 min to -0 and a -0 max to +0 (I don't think parquet-cpp did that though).

Reply:
Note that I specifically avoided fmin/fmax just to make sure that NaN would not be returned, but it looks like the c++ standard fmin/fmax behavior would be correct. Still need to double-check that cuda's fmin(NaN, x)==x, fmin(x, NaN)==x, fmax(NaN,x)==x and fmax(x, NaN)==x.

Reply:
It seems that all implementations treat +0 and -0 as equal, and I double-check in the case of parquet-cpp, there is no special case for +0 and -0, so not sure we need to do anything about it (I don't think it's possible for the writer to end up with a min=+0 and max=-0 since the comparison order is deterministic and the same for both min & max).
I'm thinking that I should switch to fmin/fmax though, since they appear to have the correct NaN behavior...

Reply:
Ok. I switched to fmin/fmax, and since the behavior of these regarding +0.0 and -0.0 isn't very well documented, I also added the explicit check to force zero min as negative zero and zero max as positive zero (presumably if someone went through the trouble of adding this in a parquet writer, there must be a good enough reason to do so, whatever that is :) ).